Human model generation and ROS support

charsyme opened the pull request:
Human model generation module for (a) generating clothed humans, (b) approximating the positions of its joints in the 3D space, (c) extracting (online) multi-images based on a list of rotations.

passalis commented:
@charsyme testing is now available again. So, when you are ready, you can use the "test sources" label to run the style checks and "test tools" to run the tool tests.

passalis commented on src/opendr/simulation/human_model_generation/utilities/visualizer.py:
```diff
@@ -0,0 +1,136 @@
+# Copyright 1996-2020 OpenDR European Project
```
This notice seems wrong. It should probably be 2020-2021

charsyme replied:
I modified it. This mistake seems to be also in face_recognition_learner.py in the master branch.

passalis reviewed (commented):
Thanks for this PR. There are a few minor issues to be addressed, mainly regarding the interface of the Learner class.

charsyme commented on src/opendr/simulation/human_model_generation/pifu_generator_learner.py:
```diff
@@ -0,0 +1,97 @@
+from src.opendr.simulation.human_model_generation.utilities.PIFu.lib.options import BaseOptions
```
Import paths are modified to the rest of the scripts, too.

charsyme commented on src/opendr/simulation/human_model_generation/pifu_generator_learner.py:
```diff
@@ -0,0 +1,97 @@
+        self.opt.norm_color = 'group'
+        self.evaluator = Evaluator(self.opt)
+
+    def infer(self, imgs_rgb, imgs_msk=None, obj_path=None, extract_pose=False):
```
Changes were made both in the module's code, as well as on the jupyter notebook demo on "projects" folder.

charsyme commented on src/opendr/simulation/human_model_generation/pifu_generator_learner.py:
```diff
@@ -0,0 +1,97 @@
+        net_C_path = os.path.join(checkpoint_dir, 'net_C')
+        if not path.exists(checkpoint_dir):
+            os.mkdir(checkpoint_dir)
+        if not path.exists(net_G_path):
```
1. download() and load() functions were added.
2. the pre-trained models use a .json file as a wrapper to be more consistent with the rest of the toolkit.

charsyme commented on src/opendr/simulation/human_model_generation/pifu_generator_learner.py:
```diff
@@ -0,0 +1,97 @@
+            print("error:", e.args)
+
+    def eval(self):
+        pass
```
Done.

stefaniapedrazzi commented:
I convert this PR to draft.
Please change the status when it is ready for review.

stefaniapedrazzi commented:
Please write a message when the PR is ready for review.

charsyme commented:
The PR is ready for review. However, there is an error with the tests of the object_tracking_3d_ab3dmot, and the PR is failing it to pass the tests.

stefaniapedrazzi commented:
If the test failure seems not to be related to the PR it is fine to request the review.
Changes and contributions can be reviewed independently from the tests result.

stefaniapedrazzi reviewed (changes requested):
You should also add a documentation page and link it in the reference index:
https://github.com/tasostefas/opendr_internal/blob/master/docs/reference/index.md

A unit test is missing for this new tool.

stefaniapedrazzi commented:
To make it work correctly, it seems you also need to add empty `__init__.py` files at least in `src/opendr/simulation` and `src/opendr/simulation/human_model_generation`.

charsyme reviewed (commented):
A documentation page (along with a link in the reference index) and a unit test have been added.

charsyme commented on tests/test_license.py:
```diff
@@ -86,13 +86,16 @@ def setUp(self):
             'src/opendr/perception/activity_recognition/datasets/utils/decoder.py',
             'projects/perception/lightweight_open_pose/jetbot/utils/pid.py',
             'src/opendr/perception/compressive_learning/multilinear_compressive_learning/algorithm/trainers.py',
+            'src/opendr/simulation/human_model_generation/pifu_generator_learner.py',
```
Code in the learner is now modified in order to be more abstract compared to the author's original implementation. Thus, I removed the learner's path from the skippedFilePaths.

charsyme commented on src/opendr/simulation/human_model_generation/pifu_generator_learner.py:
```diff
@@ -0,0 +1,132 @@
+from opendr.simulation.human_model_generation.utilities.PIFu.lib.options import BaseOptions
```
I added the licence header to the learner.

charsyme commented on src/opendr/simulation/human_model_generation/README.md:
```diff
@@ -0,0 +1,10 @@
+
+## Sources
+
+Code for [PIFu](https://arxiv.org/abs/1905.05172) is taken from [https://github.com/shunsukesaito/PIFu](https://github.com/shunsukesaito/PIFu). It is modified to be compatible with OpenDR specifications.
```
I added some details about the applied modifications.

charsyme commented:
> Thank you!
> 
> You should also add a documentation page and link it in the reference index: https://github.com/tasostefas/opendr_internal/blob/master/docs/reference/index.md
> 
> A unit test is missing for this new tool.

I added a unit test and the __init__.py files. The unit test I included doesn't seem to run at the current stage . Have I done some wrong or the test is applied after the merge of the pull-request?

ad-daniel commented on projects/opendr_ws/src/ros_bridge/src/opendr_bridge/bridge.py:
```diff
@@ -85,6 +89,34 @@ def to_ros_pose(self, pose):
             keypoints.detections.append(keypoint)
         return keypoints
 
+    def to_ros_3Dpose(self, pose):
```
If you're adding methods to the rosbridge, they should be documented in the corresponding [documentation](https://github.com/tasostefas/opendr_internal/blob/master/docs/reference/rosbridge.md)

ad-daniel commented on projects/simulation/human_model_generation/demos/model_generation.ipynb:
```diff
@@ -0,0 +1,158 @@
+{
```
It would be good to add a short README to explain how to run the jupyter notebook of the demo, what to install etc (or at least to point to where this information can be found).

charsyme commented on src/opendr/simulation/human_model_generation/pifu_generator_learner.py:
```diff
@@ -0,0 +1,134 @@
+                return [model_3D, human_poses_3D]
+            return model_3D
+        except Exception as e:
+            raise NotImplementedError('error: ' + str(e.args))
```
I think I modified the rest of those exceptions too.

charsyme commented on src/opendr/simulation/human_model_generation/pifu_generator_learner.py:
```diff
@@ -0,0 +1,134 @@
+                imgs_msk[i] = Image(imgs_msk[i])
+            imgs_msk[i] = imgs_msk[i].numpy()
+        if imgs_msk is None or len(imgs_rgb) != 1 or len(imgs_msk) != 1:
+            raise NotImplementedError('Wrong input...')
```
I changed the rest of those similar exceptions to ```ValueError``` exceptions

charsyme commented:
> Thank you, looks good to me, I only have a doubt concerning the changes you made on the bridge. The rest is minor stuff

@ad-daniel I think I addressed all the issues pointed out.

ad-daniel commented on docs/reference/rosbridge.md:
````diff
@@ -95,10 +95,101 @@ ROSBridge.from_ros_boxes(self,
+- **opendr_pose**: *engine.target.Pose*  
+  OpenDR pose to be converted to ```geometry_msgs.msg.Pose``` message.
+  
+  #### `ROSBridge.from_ros_3Dpose`
+
+```python
+ROSBridge.from_ros_pose(self,
+                        ros_pose)
+```
+
+Converts a ROS pose into an OpenDR pose (used for a 3D pose).
+
+Parameters:
+
+- **ros_pose**: *geometry_msgs.msg.Pose*  
+  ROS pose to be converted into an OpenDR pose.
+       
````
double entry, already present.

ad-daniel commented on docs/reference/rosbridge.md:
````diff
@@ -95,10 +95,101 @@ ROSBridge.from_ros_boxes(self,
+#### `ROSBridge.from_ros_3Dpose`
+
+```python
+ROSBridge.from_ros_pose(self,
````
wrong method name

ad-daniel commented on docs/reference/rosbridge.md:
```diff
@@ -95,10 +95,101 @@ ROSBridge.from_ros_boxes(self,
+Parameters:
+
+- **vertices**: *numpy.ndarray*  
+  Vertices (N) of a triangle mesh.
```
```suggestion
  Vertices (Nx3) of a triangle mesh.
```

ad-daniel replied:
According to the code, this is also Nx3.

charsyme replied:
done

charsyme commented on src/opendr/simulation/human_model_generation/pifu_generator_learner.py:
```diff
@@ -0,0 +1,134 @@
+                imgs_msk[i] = Image(imgs_msk[i])
+            imgs_msk[i] = imgs_msk[i].numpy()
+        if imgs_msk is None or len(imgs_rgb) != 1 or len(imgs_msk) != 1:
+            raise ValueError('Wrong input...')
```
Done.
